### PR TITLE
Sensor: ADXL345: Add ADXL345 driver

### PR DESCRIPTION
**Pull request by kgugala**
Add support for Digilent ADXL345 3-axis I2C accelerometer.

Signed-off-by: Kamil Rakoczy <krakoczy@antmicro.com>

### CODE REVIEW

File: drivers/sensor/adxl345/adxl345.c [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+{
+	val->val1 = sample;
+	val->val2 = 0;
+}
+
+static int adxl345_attr_set(struct device *dev, enum sensor_channel chan,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	return 0;
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: If this sensor/driver does not support setting any attributes, you should return `-ENOTSUP` in all cases.

File: drivers/sensor/adxl345/adxl345.c [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	int samples_count;
+	int rc;
+
+	data->sample_number = 0;
+	rc = adxl345_read_reg(dev, ADXL345_FIFO_STATUS_REG);
+	if (rc < 0) {
+		LOG_ERR("Failed to read FIFO status rc = %d\n", rc);
+		return rc;
+	}
+	samples_count = rc;
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: Do we need the extra `rc` variable?

File: drivers/sensor/adxl345/adxl345.c [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	data->i2c_master = device_get_binding(cfg->i2c_master_name);
+	data->i2c_addr = DT_INST_0_DIGILENT_ADXL345_BASE_ADDRESS;
+
+	if (!data->i2c_master) {
+		LOG_ERR("Failed to get I2C master\n");
+		return -EINVAL;
+	}
+
+	rc = adxl345_read_reg(dev, ADXL345_DEVICE_ID_REG);
+	id = rc;
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: The `id` variable is not used anywhere.

File: drivers/sensor/adxl345/adxl345.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+#define ADXL345_RANGE_2G           (0x0)
+#define ADXL345_RANGE_4G           (0x1)
+#define ADXL345_RANGE_8G           (0x2)
+#define ADXL345_RANGE_16G          (0x3)
+#define ADXL345_RATE_25HZ          (0x8)
+#define ADXL345_ENABLE_MEASURE_BIT (1 << 3)
+#define ADXL345_FIFO_STREAM_MODE   (1 << 7)
+#define ADXL345_FIFO_COUNT_MASK    (0x3f)
+#define ADXL345_COMPLEMENT         (0xfc00)
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.h]: Is there any reason for the parentheses?

File: drivers/sensor/adxl345/adxl345.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+struct adxl345_dev_data {
+	unsigned int sample_number;
+	struct device *i2c_master;
+	u8_t i2c_addr;
+
+	s16_t bufx[ADXL345_MAX_FIFO_SIZE];
+	s16_t bufy[ADXL345_MAX_FIFO_SIZE];
+	s16_t bufz[ADXL345_MAX_FIFO_SIZE];
+
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.h]: redundant whitespace

File: drivers/sensor/adxl345/adxl345.c [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	if (rc < 0) {
+		LOG_ERR("Enable measure bit failed\n");
+	}
+
+	return 0;
+}
+
+static int adxl345_init(struct device *dev)
+{
+	if (adxl345_probe(dev) < 0) {
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: Why is this a separate function? Can the `_probe` body be moved here?

File: drivers/sensor/adxl345/adxl345.c [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+	int rc = i2c_write_read(data->i2c_master,
+				data->i2c_addr,
+				&reg, 1,
+				&reg_val, 1);
+
+	if (rc < 0) {
+		LOG_ERR("Register read failed with rc=%d\n", rc);
+		return rc;
+	}
+	return reg_val;
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: Please add an empty line before the `return` statement at the end of a function (here and in all the other functions).

**Review by tgorochowik — CHANGES_REQUESTED** [review]

File: drivers/sensor/adxl345/adxl345.c (hunk 1 above)
Comment (tgorochowik) [reply drivers/sensor/adxl345/adxl345.c]: Or even better just remove this callback. The upper layer will return `-ENOTSUP` if this op is set to `NULL`

File: drivers/sensor/adxl345/adxl345.c (hunk 2 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: Yes, later, we are using this variable in loop to check read status.

File: drivers/sensor/adxl345/adxl345.c [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static const struct sensor_driver_api adxl345_api_funcs = {
+	.attr_set = NULL,
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: No need to do this explicitly, `adxl345_api_funcs` is static so it will be initialized to null.

**Review by tgorochowik — APPROVED** [review]
LGTM

File: drivers/sensor/adxl345/adxl345.c [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+	.sample_fetch = adxl345_sample_fetch,
+	.channel_get = adxl345_channel_get,
+};
+
+static int adxl345_init(struct device *dev)
+{
+
```
Comment (tgorochowik) [inline drivers/sensor/adxl345/adxl345.c]: Remove this empty line too.

**Review by mgielda — APPROVED** [review]
Looks good to me as well.

**Comment by mgielda** [comment]
@MaureenHelm @galak this is blocking making the integration between TensorFlow and Zephyr upstream in TF. It's very important from the promotion standpoint. We will be blogging with the TF team in the coming weeks, and it would be great to be able to include Zephyr in this.

File: drivers/sensor/adxl345/adxl345.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+	if (rc < 0) {
+		LOG_ERR("Register read failed with rc=%d\n", rc);
+		return rc;
+	}
+
+	return reg_val;
+}
+
+static int adxl345_write_reg(struct device *dev, u8_t reg, u8_t val)
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: Please use `i2c_reg_write_byte` and `i2c_reg_read_byte` directly. This function and `adxl345_read_reg` are unnecessary wrappers.

File: drivers/sensor/adxl345/adxl345.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+	if (rc < 0) {
+		LOG_ERR("Samples read failed with rc=%d\n", rc);
+		return rc;
+	}
+
+	raw_x = axis_data[0] | (axis_data[1] << 8);
+	raw_y = axis_data[2] | (axis_data[3] << 8);
+	raw_z = axis_data[4] | (axis_data[5] << 8);
+
+	if (raw_x & (1 << 9)) {
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: Please replace `(1 << 9)` with `BIT(9)` for all three axes

File: drivers/sensor/adxl345/adxl345.c [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+	sample->y = (raw_y * 1000) / 32;
+	sample->z = (raw_z * 1000) / 32;
+
+	return 0;
+}
+
+static void adxl345_accel_convert(struct sensor_value *val, unsigned int sample)
+{
+	val->val1 = sample;
+	val->val2 = 0;
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: Is `sample` already in m/s/s units?

File: drivers/sensor/adxl345/adxl345.c [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+	}
+
+	return 0;
+}
+
+static struct adxl345_dev_data adxl345_data;
+
+static const struct adxl345_dev_config adxl345_config = {
+	.data = &adxl345_data,
+	.i2c_master_name = DT_INST_0_LITEX_I2C_LABEL,
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: This should not be soc-specific. It should be `DT_INST_0_ADI_ADXL345_BUS_NAME` instead

File: drivers/sensor/adxl345/adxl345.h [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+#ifndef ZEPHYR_DRIVERS_SENSOR_ADX345_ADX345_H_
+#define ZEPHYR_DRIVERS_SENSOR_ADX345_ADX345_H_
+
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/gpio.h>
+#include <drivers/i2c.h>
+#include <sys/util.h>
+
+#define ADXL345_I2C_ADDR           0x1d
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.h]: Shouldn't need this line

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please add this sensor to `tests/drivers/build_all/sensors_a_h.conf` to ensure it gets built in CI

File: drivers/sensor/adxl345/adxl345.c (hunk 12 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: Yes, it is converted in adxl345_read_sample function.

File: drivers/sensor/adxl345/adxl345.c (hunk 10 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: Done

File: drivers/sensor/adxl345/adxl345.c (hunk 11 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: Done

File: drivers/sensor/adxl345/adxl345.c (hunk 13 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: Done

File: drivers/sensor/adxl345/adxl345.h (hunk 14 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.h]: Done

File: drivers/sensor/adxl345/Kconfig [hunk 15]
```diff
@@ -0,0 +1,10 @@
+# ADXL345, 3-Axis, +/-16g Digital Accelerometer
+
+# Copyright (c) 2020 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config ADXL345
+	bool "ADXL345 Three Axis High-g I2C accelerometer"
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/Kconfig]: This isn't a "high-g" accelerometer. Please remove the "high-g" part

File: drivers/sensor/adxl345/adxl345.c (hunk 12 above)
Comment (MaureenHelm) [reply drivers/sensor/adxl345/adxl345.c]: Please move the conversion here. It's confusing to have a `convert` function, but do the conversion somewhere else

File: drivers/sensor/adxl345/adxl345.c [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+		LOG_ERR("Rate setting failed\n");
+	}
+
+	rc = i2c_reg_write_byte(dev, data->i2c_addr, ADXL345_POWER_CTL_REG,
+			ADXL345_ENABLE_MEASURE_BIT);
+	if (rc < 0) {
+		LOG_ERR("Enable measure bit failed\n");
+	}
+
+	return 0;
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: You check for a lot of possible error conditions, but never return an error?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Please update the commit message to replace "Digilent" with "Analog Devices"

File: drivers/sensor/adxl345/Kconfig (hunk 15 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/Kconfig]: done

File: drivers/sensor/adxl345/adxl345.c (hunk 12 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: done

File: drivers/sensor/adxl345/adxl345.c (hunk 16 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: done, I added returning of error code if command fails

**Comment by kgugala** [comment]
the PR has been rebased, all the comments are addressed and CI is green. @MaureenHelm is there anything else blocking the merge?

File: drivers/sensor/adxl345/adxl345.c [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+static int adxl345_sample_fetch(struct device *dev, enum sensor_channel chan)
+{
+	struct adxl345_dev_data *data = dev->driver_data;
+	struct adxl345_sample sample;
+	u8_t samples_count;
+	int rc;
+
+	data->sample_number = 0;
+	rc = i2c_reg_read_byte(dev, data->i2c_addr, ADXL345_FIFO_STATUS_REG,
+			&samples_count);
```
Comment (MaureenHelm) [inline drivers/sensor/adxl345/adxl345.c]: check or assert that `samples_count` is less than `ADXL345_MAX_FIFO_SIZE` so you don't overrun the buffer in the loop below

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Comment by joerchan** [comment]
@kgugala I think you messed something up here.
```
git fetch <zephyr> master:master
git rebase --onto <zephyr>/master HEAD~1 adx1345
```
That should fix it. (replace <zephyr> with you zephyr remote name)

**Comment by kamilrakoczy** [comment]
Yes, it was my mistake, I'm very sorry for this. Now it should be OK.

File: drivers/sensor/adxl345/adxl345.c [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static struct adxl345_dev_data adxl345_data;
+
+static const struct adxl345_dev_config adxl345_config = {
+	.data = &adxl345_data,
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: done

File: drivers/sensor/adxl345/adxl345.c (hunk 17 above)
Comment (kamilrakoczy) [reply drivers/sensor/adxl345/adxl345.c]: done

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**